internal/query_constructor.go

```diff
@@ -0,0 +1,138 @@
+	// WorkflowStatusCanceled is the WorkflowStatus for canceled workflows
+	WorkflowStatusCanceled = WorkflowStatus(shared.WorkflowExecutionCloseStatusCanceled.String())
+	// WorkflowStatusTerminated is the WorkflowStatus for terminated workflows
+	WorkflowStatusTerminated = WorkflowStatus(shared.WorkflowExecutionCloseStatusTerminated.String())
+	// WorkflowStatusContinuedAsNew is the WorkflowStatus for continuedAsNew workflows
+	WorkflowStatusContinuedAsNew = WorkflowStatus(shared.WorkflowExecutionCloseStatusContinuedAsNew.String())
+	// WorkflowStatusTimedOut is the WorkflowStatus for timedout workflows
+	WorkflowStatusTimedOut = WorkflowStatus(shared.WorkflowExecutionCloseStatusTimedOut.String())
+)
+
+const (
+	keyWorkflowType = "WorkflowType"
+	keyCloseStatus  = "CloseStatus"
+	keyStartTime    = "StartTime"
+	keyCloseTime    = "CloseTime"
+)
+
+var (
+	maxTimestamp = time.Unix(0, math.MaxInt64)
+)
+
+type (
+	// QueryConstructor constructs visibility query
+	QueryConstructor interface {
+		WorkflowTypes([]string) QueryConstructor
+		WorkflowStatus([]WorkflowStatus) QueryConstructor
+		StartTime(time.Time, time.Time) QueryConstructor
+		Query() string
+	}
+
+	queryConstructorImpl struct {
+		builder strings.Builder
+	}
+)
+
+// NewQueryConstructor creates a new visibility QueryConstructor
+func NewQueryConstructor() QueryConstructor {
+	return &queryConstructorImpl{}
+}
+
+func (q *queryConstructorImpl) WorkflowTypes(types []string) QueryConstructor {
+	workflowTypeQueries := make([]string, 0, len(types))
+	for _, workflowType := range types {
+		workflowTypeQueries = append(workflowTypeQueries, fmt.Sprintf(keyWorkflowType+` = "%v"`, workflowType))
+	}
+	q.appendPartialQuery(strings.Join(workflowTypeQueries, " or "))
+	return q
+}
+
+func (q *queryConstructorImpl) WorkflowStatus(statuses []WorkflowStatus) QueryConstructor {
+	workflowStatusQueries := make([]string, 0, len(statuses))
+	for _, status := range statuses {
+		var statusQuery string
+		switch status {
+		case WorkflowStatusOpen:
+			statusQuery = keyCloseTime + " = missing"
+		case WorkflowStatusClosed:
+			statusQuery = keyCloseTime + " != missing"
+		default:
+			statusQuery = keyCloseStatus + ` = "` + string(status) + `"`
+		}
+		workflowStatusQueries = append(workflowStatusQueries, statusQuery)
+	}
+	q.appendPartialQuery(strings.Join(workflowStatusQueries, " or "))
+	return q
+}
+
+func (q *queryConstructorImpl) StartTime(minStartTime, maxStartTime time.Time) QueryConstructor {
+	startTimeQueries := make([]string, 0, 2)
+	if !minStartTime.IsZero() {
+		startTimeQueries = append(startTimeQueries, fmt.Sprintf(keyStartTime+` >= %v`, minStartTime.UnixNano()))
+	}
+	if !maxStartTime.Equal(maxTimestamp) {
+		startTimeQueries = append(startTimeQueries, fmt.Sprintf(keyStartTime+` <= %v`, maxStartTime.UnixNano()))
+	}
+
+	q.appendPartialQuery(strings.Join(startTimeQueries, " and "))
+	return q
+}
+
+func (q *queryConstructorImpl) Query() string {
```

Comment:
maybe named this query builder and build() function?

Reply:
Makes sense. Will update.

Reply:
Renamed to `QueryBuilder` and `Build()`